#include "addrecipedialog.h"
#include "ui_addrecipedialog.h"
#include <QFileDialog>
#include <QMessageBox>
#include <QSqlQuery>
#include <QSqlError>
#include <QPixmap>
#include <QFileInfo>
#include <QDir>
#include <QDateTime>

AddRecipeDialog::AddRecipeDialog(int userId, QWidget *parent) :
    QDialog(parent),
    ui(new Ui::AddRecipeDialog),
    currentUserId(userId),
    currentRecipeId(0),
    photoPath(""),
    currentStepNumber(1),
    photoDialogOpen(false)
{
    ui->setupUi(this);
    loadCategories();

    // –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –µ–¥–∏–Ω–∏—Ü –∏–∑–º–µ—Ä–µ–Ω–∏—è
    ui->ingredientUnitComboBox->addItems({"–µ–¥.", "–≥", "–∫–≥", "–º–ª", "–ª", "—à—Ç", "—á.–ª.", "—Å—Ç.–ª.", "–ø–æ –≤–∫—É—Å—É"});
    ui->ingredientUnitComboBox->setCurrentIndex(0);

    // –ê–≤—Ç–æ–º–∞—Ç–∏—á–µ—Å–∫–∞—è –Ω—É–º–µ—Ä–∞—Ü–∏—è —à–∞–≥–æ–≤
    updateCurrentStepLabel();

    // –£–¥–∞–ª–µ–Ω–∏–µ –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç–æ–≤ –∏ —à–∞–≥–æ–≤ –¥–≤–æ–π–Ω—ã–º –∫–ª–∏–∫–æ–º
    connect(ui->ingredientsListWidget, &QListWidget::itemDoubleClicked, this, &AddRecipeDialog::removeIngredient);
    connect(ui->stepsListWidget, &QListWidget::itemDoubleClicked, this, &AddRecipeDialog::removeStep);
    
    // –í—ã–±–æ—Ä —Ñ–æ—Ç–æ
    //connect(ui->addPhotoButton, &QPushButton::clicked, this, &AddRecipeDialog::on_addPhotoButton_clicked);
    
    // –î–µ–ª–∞–µ–º photoLabel –∫–ª–∏–∫–∞–±–µ–ª—å–Ω—ã–º
    ui->photoLabel->setCursor(Qt::PointingHandCursor);
    ui->photoLabel->installEventFilter(this);
    
    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –Ω–∞—á–∞–ª—å–Ω–æ–µ –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ –¥–ª—è —Ñ–æ—Ç–æ
    updatePhotoDisplay();
}

AddRecipeDialog::AddRecipeDialog(int userId, int recipeId, QWidget *parent) :
    QDialog(parent),
    ui(new Ui::AddRecipeDialog),
    currentUserId(userId),
    currentRecipeId(recipeId),
    photoPath(""),
    currentStepNumber(1)
{
    ui->setupUi(this);
    loadCategories();

    // –ò–Ω–∏—Ü–∏–∞–ª–∏–∑–∞—Ü–∏—è –µ–¥–∏–Ω–∏—Ü –∏–∑–º–µ—Ä–µ–Ω–∏—è
    ui->ingredientUnitComboBox->addItems({"–µ–¥.", "–≥", "–∫–≥", "–º–ª", "–ª", "—à—Ç", "—á.–ª.", "—Å—Ç.–ª.", "–ø–æ –≤–∫—É—Å—É"});
    ui->ingredientUnitComboBox->setCurrentIndex(0);

    // –ó–∞–≥—Ä—É–∂–∞–µ–º –¥–∞–Ω–Ω—ã–µ —Ä–µ—Ü–µ–ø—Ç–∞ –¥–ª—è —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏—è
    loadRecipeData();

    // –£–¥–∞–ª–µ–Ω–∏–µ –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç–æ–≤ –∏ —à–∞–≥–æ–≤ –¥–≤–æ–π–Ω—ã–º –∫–ª–∏–∫–æ–º
    connect(ui->ingredientsListWidget, &QListWidget::itemDoubleClicked, this, &AddRecipeDialog::removeIngredient);
    connect(ui->stepsListWidget, &QListWidget::itemDoubleClicked, this, &AddRecipeDialog::removeStep);
    
    // –í—ã–±–æ—Ä —Ñ–æ—Ç–æ
    connect(ui->addPhotoButton, &QPushButton::clicked, this, &AddRecipeDialog::on_addPhotoButton_clicked);
    
    // –î–µ–ª–∞–µ–º photoLabel –∫–ª–∏–∫–∞–±–µ–ª—å–Ω—ã–º
    ui->photoLabel->setCursor(Qt::PointingHandCursor);
    ui->photoLabel->installEventFilter(this);
    
    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –Ω–∞—á–∞–ª—å–Ω–æ–µ –∏–∑–æ–±—Ä–∞–∂–µ–Ω–∏–µ –¥–ª—è —Ñ–æ—Ç–æ
    updatePhotoDisplay();
    
    // –ú–µ–Ω—è–µ–º —Ç–µ–∫—Å—Ç –∫–Ω–æ–ø–∫–∏ –Ω–∞ "–°–æ—Ö—Ä–∞–Ω–∏—Ç—å"
    ui->saveButton->setText("–°–æ—Ö—Ä–∞–Ω–∏—Ç—å");
}

AddRecipeDialog::~AddRecipeDialog()
{
    delete ui;
}

bool AddRecipeDialog::eventFilter(QObject *obj, QEvent *event)
{
    if (event->type() == QEvent::MouseButtonPress) {
        // –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º –∫–ª–∏–∫ —Ç–æ–ª—å–∫–æ –ø–æ photoLabel
        if (obj == ui->photoLabel) {
            on_addPhotoButton_clicked();
            return true;
        }
    }
    return QDialog::eventFilter(obj, event);
}

void AddRecipeDialog::loadCategories()
{
    ui->categoryComboBox->clear();
    ui->categoryComboBox->addItem("–í—ã–±–µ—Ä–∏—Ç–µ –∏–∑ —Å–ø–∏—Å–∫–∞...", QVariant()); // –ü—É—Å—Ç–æ–π —ç–ª–µ–º–µ–Ω—Ç –¥–ª—è placeholder
    QSqlQuery query("SELECT name FROM categories ORDER BY name");
    while (query.next()) {
        ui->categoryComboBox->addItem(query.value(0).toString());
    }
    ui->categoryComboBox->setCurrentIndex(0); // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –Ω–∞ placeholder
}

void AddRecipeDialog::updateCurrentStepLabel()
{
    ui->currentStepLabel->setText(QString("–®–ê–ì %1:").arg(currentStepNumber));
}

// void AddRecipeDialog::on_addPhotoButton_clicked()
// {
//     // –ë–ª–æ–∫–∏—Ä—É–µ–º –ø–æ–≤—Ç–æ—Ä–Ω–æ–µ –æ—Ç–∫—Ä—ã—Ç–∏–µ –¥–∏–∞–ª–æ–≥–∞, –µ—Å–ª–∏ –æ–Ω —É–∂–µ –æ—Ç–∫—Ä—ã—Ç
//     static bool dialogOpen = false;
//     if (dialogOpen) {
//         return;
//     }
    
//     dialogOpen = true;
//     QString fileName = QFileDialog::getOpenFileName(this,
//         tr("–í—ã–±–µ—Ä–∏—Ç–µ —Ñ–æ—Ç–æ"), "",
//         tr("–ò–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è (*.png *.jpg *.jpeg *.bmp);;–í—Å–µ —Ñ–∞–π–ª—ã (*)"));
//     dialogOpen = false;
    
//     if (!fileName.isEmpty()) {
//         photoPath = fileName;
//         updatePhotoDisplay();
//     }
// }

void AddRecipeDialog::on_addPhotoButton_clicked()
{
    // –ü—Ä–µ–¥–æ—Ç–≤—Ä–∞—â–∞–µ–º –ø–æ–≤—Ç–æ—Ä–Ω–æ–µ –æ—Ç–∫—Ä—ã—Ç–∏–µ –¥–∏–∞–ª–æ–≥–∞
    if (photoDialogOpen) {
        return;
    }

    photoDialogOpen = true;

    QString fileName = QFileDialog::getOpenFileName(this,
        tr("–í—ã–±–µ—Ä–∏—Ç–µ —Ñ–æ—Ç–æ"), "",
        tr("–ò–∑–æ–±—Ä–∞–∂–µ–Ω–∏—è (*.png *.jpg *.jpeg *.bmp);;–í—Å–µ —Ñ–∞–π–ª—ã (*)"));

    photoDialogOpen = false;

    if (!fileName.isEmpty()) {
        photoPath = fileName;
        updatePhotoDisplay();
    }
}

void AddRecipeDialog::updatePhotoDisplay()
{
    if (!photoPath.isEmpty() && QFile::exists(photoPath)) {
        QPixmap pix(photoPath);
        ui->photoLabel->setPixmap(pix.scaled(150, 150, Qt::KeepAspectRatioByExpanding, Qt::SmoothTransformation));
        ui->photoLabel->setStyleSheet("border: 2px solid #ced4da; border-radius: 20px;");
    } else {
        // –ü–æ–∫–∞–∑—ã–≤–∞–µ–º –∏–∫–æ–Ω–∫—É –ø–ª—é—Å–∞ –Ω–∞ —Ñ–æ—Ç–æ
        ui->photoLabel->setText("üì∑");
        ui->photoLabel->setStyleSheet("border: 2px dashed #ced4da; border-radius: 20px; background-color: #f8f9fa; font-size: 60px;");
    }
}

void AddRecipeDialog::loadRecipeData()
{
    if (currentRecipeId <= 0) {
        return;
    }
    
    QSqlQuery query;
    query.prepare("SELECT name, time, photo_path, category_id FROM recipes WHERE id = ? AND user_id = ?");
    query.addBindValue(currentRecipeId);
    query.addBindValue(currentUserId);
    
    if (!query.exec() || !query.next()) {
        QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–†–µ—Ü–µ–ø—Ç –Ω–µ –Ω–∞–π–¥–µ–Ω");
        return;
    }
    
    QString name = query.value(0).toString();
    int time = query.value(1).toInt();
    QString savedPhotoPath = query.value(2).toString();
    int categoryId = query.value(3).toInt();
    
    // –ó–∞–ø–æ–ª–Ω—è–µ–º –ø–æ–ª—è
    ui->nameLineEdit->setText(name);
    ui->timeSpinBox->setValue(time);
    
    // –£—Å—Ç–∞–Ω–∞–≤–ª–∏–≤–∞–µ–º –∫–∞—Ç–µ–≥–æ—Ä–∏—é
    QSqlQuery catQuery;
    catQuery.prepare("SELECT name FROM categories WHERE id = ?");
    catQuery.addBindValue(categoryId);
    if (catQuery.exec() && catQuery.next()) {
        QString categoryName = catQuery.value(0).toString();
        int index = ui->categoryComboBox->findText(categoryName);
        if (index >= 0) {
            ui->categoryComboBox->setCurrentIndex(index);
        }
    }
    
    // –ó–∞–≥—Ä—É–∂–∞–µ–º —Ñ–æ—Ç–æ
    if (!savedPhotoPath.isEmpty()) {
        photoPath = savedPhotoPath;
    }
    
    // –ó–∞–≥—Ä—É–∂–∞–µ–º –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç—ã
    QSqlQuery ingQuery;
    ingQuery.prepare("SELECT name, quantity FROM ingredients WHERE recipe_id = ? ORDER BY id");
    ingQuery.addBindValue(currentRecipeId);
    ingQuery.exec();
    
    while (ingQuery.next()) {
        QString name = ingQuery.value(0).toString();
        QString quantity = ingQuery.value(1).toString();
        
        QString itemText = name;
        if (!quantity.isEmpty() && quantity != "–ø–æ –≤–∫—É—Å—É") {
            itemText += " ‚Äî " + quantity;
        }
        
        ui->ingredientsListWidget->addItem(itemText);
    }
    
    // –ó–∞–≥—Ä—É–∂–∞–µ–º —à–∞–≥–∏
    QSqlQuery stepsQuery;
    stepsQuery.prepare("SELECT step_number, description FROM steps WHERE recipe_id = ? ORDER BY step_number");
    stepsQuery.addBindValue(currentRecipeId);
    stepsQuery.exec();
    
    while (stepsQuery.next()) {
        QString desc = stepsQuery.value(1).toString();
        ui->stepsListWidget->addItem(QString("%1. %2").arg(currentStepNumber).arg(desc));
        currentStepNumber++;
    }
    updateCurrentStepLabel();
}

void AddRecipeDialog::on_backButton_clicked()
{
    reject();
}

void AddRecipeDialog::on_addIngredientButton_clicked()
{
    QString name = ui->ingredientNameEdit->text().trimmed();
    int quantity = ui->ingredientQuantitySpinBox->value();
    QString unit = ui->ingredientUnitComboBox->currentText();

    if (name.isEmpty()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –Ω–∞–∑–≤–∞–Ω–∏–µ –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç–∞");
        return;
    }

    QString itemText = name;
    if (quantity > 0) {
        itemText += QString(" ‚Äî %1 %2").arg(quantity).arg(unit);
    } else if (!unit.isEmpty() && unit != "–ø–æ –≤–∫—É—Å—É" && unit != "–µ–¥.") {
        itemText += QString(" ‚Äî %1").arg(unit);
    }

    ui->ingredientsListWidget->addItem(itemText);

    ui->ingredientNameEdit->clear();
    ui->ingredientQuantitySpinBox->setValue(0);
    ui->ingredientUnitComboBox->setCurrentIndex(0);
}

void AddRecipeDialog::removeIngredient(QListWidgetItem *item)
{
    delete ui->ingredientsListWidget->takeItem(ui->ingredientsListWidget->row(item));
}

void AddRecipeDialog::on_addStepButton_clicked()
{
    QString desc = ui->stepDescriptionEdit->toPlainText().trimmed();
    if (desc.isEmpty()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –æ–ø–∏—Å–∞–Ω–∏–µ —à–∞–≥–∞");
        return;
    }

    QString itemText = QString("%1. %2").arg(currentStepNumber).arg(desc);
    ui->stepsListWidget->addItem(itemText);

    currentStepNumber++;
    updateCurrentStepLabel();
    ui->stepDescriptionEdit->clear();
}

void AddRecipeDialog::removeStep(QListWidgetItem *item)
{
    int row = ui->stepsListWidget->row(item);
    delete ui->stepsListWidget->takeItem(row);

    // –ü–µ—Ä–µ–Ω—É–º–µ—Ä–æ–≤—ã–≤–∞–µ–º –æ—Å—Ç–∞–≤—à–∏–µ—Å—è —à–∞–≥–∏
    for (int i = row; i < ui->stepsListWidget->count(); ++i) {
        QString text = ui->stepsListWidget->item(i)->text();
        text = text.section(". ", 1);
        ui->stepsListWidget->item(i)->setText(QString("%1. %2").arg(i + 1).arg(text));
    }
    currentStepNumber = ui->stepsListWidget->count() + 1;
    updateCurrentStepLabel();
}

void AddRecipeDialog::on_saveButton_clicked()
{
    QString name = ui->nameLineEdit->text().trimmed();
    int time = ui->timeSpinBox->value();
    QString categoryName = ui->categoryComboBox->currentText();

    // –í–∞–ª–∏–¥–∞—Ü–∏—è –æ–±—è–∑–∞—Ç–µ–ª—å–Ω—ã—Ö –ø–æ–ª–µ–π
    if (name.isEmpty()) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í–≤–µ–¥–∏—Ç–µ –Ω–∞–∑–≤–∞–Ω–∏–µ –±–ª—é–¥–∞");
        ui->nameLineEdit->setFocus();
        return;
    }
    
    if (categoryName.isEmpty() || categoryName == "–í—ã–±–µ—Ä–∏—Ç–µ –∏–∑ —Å–ø–∏—Å–∫–∞..." || ui->categoryComboBox->currentIndex() <= 0) {
        QMessageBox::warning(this, "–û—à–∏–±–∫–∞", "–í—ã–±–µ—Ä–∏—Ç–µ –∫–∞—Ç–µ–≥–æ—Ä–∏—é –±–ª—é–¥–∞");
        ui->categoryComboBox->setFocus();
        return;
    }
    
    // –ü–æ–ª—É—á–∞–µ–º ID –∫–∞—Ç–µ–≥–æ—Ä–∏–∏ –ø–æ –∏–º–µ–Ω–∏
    QSqlQuery catQuery;
    catQuery.prepare("SELECT id FROM categories WHERE name = ?");
    catQuery.addBindValue(categoryName);
    if (!catQuery.exec() || !catQuery.next()) {
        QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –Ω–∞–π—Ç–∏ –∫–∞—Ç–µ–≥–æ—Ä–∏—é");
        return;
    }
    int categoryId = catQuery.value(0).toInt();

    // –û–±—Ä–∞–±–∞—Ç—ã–≤–∞–µ–º —Ñ–æ—Ç–æ
    QString savedPhotoPath = "";
    if (currentRecipeId > 0) {
        // –ü—Ä–∏ —Ä–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏–∏: –µ—Å–ª–∏ —Ñ–æ—Ç–æ –Ω–µ –∏–∑–º–µ–Ω–∏–ª–æ—Å—å, –∏—Å–ø–æ–ª—å–∑—É–µ–º —Å—Ç–∞—Ä–æ–µ
        QSqlQuery oldPhotoQuery;
        oldPhotoQuery.prepare("SELECT photo_path FROM recipes WHERE id = ?");
        oldPhotoQuery.addBindValue(currentRecipeId);
        if (oldPhotoQuery.exec() && oldPhotoQuery.next()) {
            QString oldPhotoPath = oldPhotoQuery.value(0).toString();
            // –ü—Ä–æ–≤–µ—Ä—è–µ–º, –∏–∑–º–µ–Ω–∏–ª–æ—Å—å –ª–∏ —Ñ–æ—Ç–æ
            if (!photoPath.isEmpty() && photoPath != oldPhotoPath && QFile::exists(photoPath)) {
                // –§–æ—Ç–æ –∏–∑–º–µ–Ω–∏–ª–æ—Å—å - –∫–æ–ø–∏—Ä—É–µ–º –Ω–æ–≤–æ–µ
                QDir dir;
                if (!dir.exists("photos")) {
                    dir.mkpath("photos");
                }
                QFileInfo fileInfo(photoPath);
                QString newFileName = QString("photos/recipe_%1_%2.%3")
                    .arg(currentUserId)
                    .arg(QDateTime::currentDateTime().toSecsSinceEpoch())
                    .arg(fileInfo.suffix());
                
                if (QFile::copy(photoPath, newFileName)) {
                    savedPhotoPath = newFileName;
                } else {
                    savedPhotoPath = photoPath;
                }
            } else {
                // –§–æ—Ç–æ –Ω–µ –∏–∑–º–µ–Ω–∏–ª–æ—Å—å - –∏—Å–ø–æ–ª—å–∑—É–µ–º —Å—Ç–∞—Ä–æ–µ
                savedPhotoPath = oldPhotoPath;
            }
        }
    } else {
        // –ù–æ–≤—ã–π —Ä–µ—Ü–µ–ø—Ç - –∫–æ–ø–∏—Ä—É–µ–º —Ñ–æ—Ç–æ
        if (!photoPath.isEmpty() && QFile::exists(photoPath)) {
            QDir dir;
            if (!dir.exists("photos")) {
                dir.mkpath("photos");
            }
            QFileInfo fileInfo(photoPath);
            QString newFileName = QString("photos/recipe_%1_%2.%3")
                .arg(currentUserId)
                .arg(QDateTime::currentDateTime().toSecsSinceEpoch())
                .arg(fileInfo.suffix());
            
            if (QFile::copy(photoPath, newFileName)) {
                savedPhotoPath = newFileName;
            } else {
                savedPhotoPath = photoPath;
            }
        }
    }

    // –ù–∞—á–∏–Ω–∞–µ–º —Ç—Ä–∞–Ω–∑–∞–∫—Ü–∏—é –¥–ª—è —Ü–µ–ª–æ—Å—Ç–Ω–æ—Å—Ç–∏ –¥–∞–Ω–Ω—ã—Ö
    QSqlDatabase db = QSqlDatabase::database();
    db.transaction();

    QSqlQuery query;
    int recipeId = currentRecipeId;
    
    if (currentRecipeId > 0) {
        // –†–µ–¥–∞–∫—Ç–∏—Ä–æ–≤–∞–Ω–∏–µ —Å—É—â–µ—Å—Ç–≤—É—é—â–µ–≥–æ —Ä–µ—Ü–µ–ø—Ç–∞
        query.prepare("UPDATE recipes SET name = ?, photo_path = ?, time = ?, category_id = ? WHERE id = ? AND user_id = ?");
        query.addBindValue(name);
        query.addBindValue(savedPhotoPath);
        query.addBindValue(time);
        query.addBindValue(categoryId);
        query.addBindValue(currentRecipeId);
        query.addBindValue(currentUserId);
        
        if (!query.exec()) {
            db.rollback();
            QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å –æ–±–Ω–æ–≤–∏—Ç—å —Ä–µ—Ü–µ–ø—Ç: " + query.lastError().text());
            return;
        }
        
        // –£–¥–∞–ª—è–µ–º —Å—Ç–∞—Ä—ã–µ –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç—ã –∏ —à–∞–≥–∏
        query.prepare("DELETE FROM ingredients WHERE recipe_id = ?");
        query.addBindValue(currentRecipeId);
        query.exec();
        
        query.prepare("DELETE FROM steps WHERE recipe_id = ?");
        query.addBindValue(currentRecipeId);
        query.exec();
    } else {
        // –î–æ–±–∞–≤–ª–µ–Ω–∏–µ –Ω–æ–≤–æ–≥–æ —Ä–µ—Ü–µ–ø—Ç–∞
        query.prepare("INSERT INTO recipes (user_id, name, photo_path, time, category_id) VALUES (?, ?, ?, ?, ?)");
        query.addBindValue(currentUserId);
        query.addBindValue(name);
        query.addBindValue(savedPhotoPath);
        query.addBindValue(time);
        query.addBindValue(categoryId);
        if (!query.exec()) {
            db.rollback();
            QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å —Ä–µ—Ü–µ–ø—Ç: " + query.lastError().text());
            return;
        }
        
        recipeId = query.lastInsertId().toInt();
    }

    // –°–æ—Ö—Ä–∞–Ω—è–µ–º –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç—ã
    for (int i = 0; i < ui->ingredientsListWidget->count(); ++i) {
        QString text = ui->ingredientsListWidget->item(i)->text();
        QString ingName = text.section(" ‚Äî ", 0, 0);
        QString quantity = text.section(" ‚Äî ", 1);
        if (quantity.isEmpty()) {
            quantity = "–ø–æ –≤–∫—É—Å—É";
        }

        query.prepare("INSERT INTO ingredients (recipe_id, name, quantity) VALUES (?, ?, ?)");
        query.addBindValue(recipeId);
        query.addBindValue(ingName);
        query.addBindValue(quantity);
        if (!query.exec()) {
            db.rollback();
            QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å –∏–Ω–≥—Ä–µ–¥–∏–µ–Ω—Ç: " + query.lastError().text());
            return;
        }
    }

    // –°–æ—Ö—Ä–∞–Ω—è–µ–º —à–∞–≥–∏
    for (int i = 0; i < ui->stepsListWidget->count(); ++i) {
        QString text = ui->stepsListWidget->item(i)->text();
        QString desc = text.section(". ", 1);

        query.prepare("INSERT INTO steps (recipe_id, step_number, description) VALUES (?, ?, ?)");
        query.addBindValue(recipeId);
        query.addBindValue(i + 1);
        query.addBindValue(desc);
        if (!query.exec()) {
            db.rollback();
            QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å —à–∞–≥: " + query.lastError().text());
            return;
        }
    }

    // –ö–æ–º–º–∏—Ç–∏–º —Ç—Ä–∞–Ω–∑–∞–∫—Ü–∏—é
    if (!db.commit()) {
        db.rollback();
        QMessageBox::critical(this, "–û—à–∏–±–∫–∞", "–ù–µ —É–¥–∞–ª–æ—Å—å —Å–æ—Ö—Ä–∞–Ω–∏—Ç—å –¥–∞–Ω–Ω—ã–µ –≤ –±–∞–∑—É");
        return;
    }

    accept(); // –£—Å–ø–µ—à–Ω–æ –∑–∞–∫—Ä—ã–≤–∞–µ–º –¥–∏–∞–ª–æ–≥
}
